chain/epoch-manager/src/lib.rs

```diff
@@ -1900,6 +1901,81 @@ impl EpochManager {
         }
     }
 
+    pub fn possible_epochs_of_height_around_tip(
+        &self,
+        tip: &Tip,
+        height: BlockHeight,
+    ) -> Result<Vec<EpochId>, EpochError> {
+        // If the tip is at the genesis block, it has to be handled in a special way.
+        // For genesis block, epoch_first_block() is the dummy block (11111...)
+        // with height 0, which could cause issues with estimating the epoch end
+        // if the genesis height is nonzero. It's easier to handle it manually.
+        if tip.prev_block_hash == CryptoHash::default() {
+            if tip.height == height {
+                return Ok(vec![tip.epoch_id.clone()]);
+            }
+
+            if height > tip.height {
+                return Ok(vec![tip.next_epoch_id.clone()]);
+            }
+
+            return Ok(vec![]);
+        }
+
+        // See if the height is in the current epoch
+        let current_epoch_first_block_hash =
+            *self.get_block_info(&tip.last_block_hash)?.epoch_first_block();
+        let current_epoch_first_block_info =
+            self.get_block_info(&current_epoch_first_block_hash)?;
+
+        let current_epoch_start = current_epoch_first_block_info.height();
+        let current_epoch_length = self.get_epoch_config(&tip.epoch_id)?.epoch_length;
+        let current_epoch_estimated_end = current_epoch_start.saturating_add(current_epoch_length);
+
+        // All blocks with height lower than the estimated end are guaranteed to reside in the current epoch.
+        // The situation is clear here.
+        if (current_epoch_start..current_epoch_estimated_end).contains(&height) {
+            return Ok(vec![tip.epoch_id.clone()]);
+        }
+
+        // If the height is higher than the current epoch's estimated end, then it's
+        // not clear in which epoch it'll be. Under normal circumstances it would be
+        // in the next epoch, but with missing blocks the current epoch could stretch out
+        // past its estimated end, so the height might end up being in the current epoch,
+        // even though its height is higher than the estimated end.
+        if height >= current_epoch_estimated_end {
+            return Ok(vec![tip.epoch_id.clone(), tip.next_epoch_id.clone()]);
+        }
+
+        // Finally try the previous epoch.
+        // First and last blocks of the previous epoch are already known, so the situation is clear.
```

Comment:
In theory we could, but it wasn't needed for orphan witnesses, so I didn't implement it.
It's also safer to validate only the epochs around the tip. A malicious actor could submit a witness at height 0 and that'd force us to walk the whole blockchain to find out its epoch

Reply:
I would expect that we already have the height -> epoch id mapping stored in the db in one form or another so we wouldn't need to actually walk the chain. But Longarithm's point is even better.

Reply:
> it actually means that we never process it because heights of processed blocks only increase.

Is that the case? I thought that sometimes we process blocks which are behind the current head.
For exampel:
* There's block `#1`
* Node X produces block `#2`, but it doesn't reach the rest of the network
* Block producers emit Skip(1, 3)
* Node Y produces block `#3`
* Y shares the block with some peers, but not all of them before ...
* Block `#2` arrives at the block producers, who vote to accept it
* Block `#4` is built on top of block `#2`, block `#3` is rejected

In that case some nodes would process block `#3` and then `#2`, and the heights would be decreasing.

Reply:
I couldn't find a scenario where we process an orphaned witness with height less then the height of the head, but I'm a noob at doomslug, so maybe it's also possible somehow?
I would leave the logic that tries the previous epoch as well, just to be safe.